File: Data Structures & Algorithms/count-paths/submission-31.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        def dfs(i, j):  
            if i == 0 or j == 0: 
                return 0 

            if i == 1 or j == 1: 
                return 1 

            return dfs(i - 1, j) + dfs(i, j - 1) 

        return dfs(m, n)

File: Data Structures & Algorithms/combination-target-sum-ii/submission-31.js
class Solution {
    /**
     * @param {number[]} candidates
     * @param {number} target
     * @return {number[][]}
     */
    combinationSum2(candidates, target) { 
        candidates.sort((a, b) => a - b) 
        const res = []
        function dfs(start, part) {
            if (target === 0) {
                res.push([...part]) 
                return
            }  

            if (target < 0) return 

            for (let i = start; i < candidates.length; i ++) { 
                if (i > start && candidates[i] === candidates[i - 1]) continue
                part.push(candidates[i]) 
                target -= candidates[i] 
                dfs(i + 1, part) 
                part.pop() 
                target += candidates[i]
            }
        } 

        dfs(0, []) 

        return res
    }
}
